File: ios/ICEInfo/Network/OsmRepository.swift
import Foundation

actor OsmRepository {
    static let shared = OsmRepository()

    private let overpassUrl = "https://overpass-api.de/api/interpreter"
    private let lookaheadDeg = 0.18
    private let trackRadiusM = 200

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        return d
    }()

    func fetchTrackData(lat: Double, lon: Double) async -> OsmTrackData {
        guard lat != 0 || lon != 0 else { return OsmTrackData(error: nil) }
        do {
            async let trackInfo = fetchTrackInfo(lat: lat, lon: lon)
            async let features = fetchFeatures(lat: lat, lon: lon)
            return try await OsmTrackData(trackInfo: trackInfo, features: features)
        } catch {
            return OsmTrackData(error: "Streckendaten konnten nicht geladen werden.")
        }
    }

    private func fetchTrackInfo(lat: Double, lon: Double) async -> TrackInfo {
        let query = "[out:json][timeout:15];way(around:\(trackRadiusM),\(lat),\(lon))[\"railway\"=\"rail\"];out tags;"
        guard let url = URL(string: overpassUrl) else { return TrackInfo() }
        var request = URLRequest(url: url, timeoutInterval: 20)
        request.httpMethod = "POST"
        request.httpBody = "data=\(query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query)".data(using: .utf8)
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let response = try decoder.decode(OverpassResponse.self, from: data)
            guard let tags = response.elements.first?.tags else { return TrackInfo() }
            return TrackInfo(
                maxSpeed: Int(tags["maxspeed"] ?? ""),
                electrified: tags["electrified"],
                voltage: Int(tags["voltage"] ?? ""),
                tracks: Int(tags["tracks"] ?? ""),
                usage: tags["usage"]
            )
        } catch {
            return TrackInfo()
        }
    }

    private func fetchFeatures(lat: Double, lon: Double) async -> [RailFeature] {
        let south = lat - lookaheadDeg
        let north = lat + lookaheadDeg
        let west = lon - lookaheadDeg
        let east = lon + lookaheadDeg
        let bbox = "\(south),\(west),\(north),\(east)"

        let query = "[out:json][timeout:20];(way[\"railway\"=\"rail\"][\"tunnel\"=\"yes\"](\(bbox));way[\"railway\"=\"rail\"][\"bridge\"=\"yes\"](\(bbox));node[\"railway\"=\"station\"](\(bbox));node[\"railway\"=\"halt\"](\(bbox)););out tags center;"

        guard let url = URL(string: overpassUrl) else { return [] }
        var request = URLRequest(url: url, timeoutInterval: 25)
        request.httpMethod = "POST"
        request.httpBody = "data=\(query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query)".data(using: .utf8)
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let response = try decoder.decode(OverpassResponse.self, from: data)
            return response.elements.compactMap { element -> RailFeature? in
                let elLat = element.lat ?? element.center?.lat ?? 0
                let elLon = element.lon ?? element.center?.lon ?? 0
                guard elLat != 0, elLon != 0 else { return nil }
                let tags = element.tags

                let type: RailFeatureType
                if tags["railway"] == "station" { type = .station }
                else if tags["railway"] == "halt" { type = .halt }
                else if tags["tunnel"] == "yes" { type = .tunnel }
                else if tags["bridge"] == "yes" { type = .bridge }
                else { return nil }

                let name = tags["name"] ?? tags["tunnel:name"] ?? tags["bridge:name"]
                if (type == .bridge || type == .station || type == .halt) && (name == nil || name!.isEmpty) { return nil }

                return RailFeature(
                    name: name ?? "Tunnel",
                    type: type,
                    distanceKm: haversineKm(lat1: lat, lon1: lon, lat2: elLat, lon2: elLon)
                )
            }
            .sorted { $0.distanceKm < $1.distanceKm }
            .reduce([]) { acc, feature in
                if acc.contains(where: { $0.type == feature.type && $0.name == feature.name }) { return acc }
                return acc + [feature]
            }
            .prefix(10)
            .map { $0 }
        } catch {
            return []
        }
    }

    private func haversineKm(lat1: Double, lon1: Double, lat2: Double, lon2: Double) -> Double {
        let r = 6371.0
        let dLat = (lat2 - lat1) * .pi / 180
        let dLon = (lon2 - lon1) * .pi / 180
        let a = sin(dLat / 2) * sin(dLat / 2) +
                cos(lat1 * .pi / 180) * cos(lat2 * .pi / 180) * sin(dLon / 2) * sin(dLon / 2)
        return r * 2 * atan2(sqrt(a), sqrt(1 - a))
    }
}

File: app/src/main/java/com/nruge/iceinfo/OsmRepository.kt
package com.nruge.iceinfo

import android.util.Log
import com.nruge.iceinfo.model.OsmTrackData
import com.nruge.iceinfo.model.RailFeature
import com.nruge.iceinfo.model.RailFeatureType
import com.nruge.iceinfo.model.TrackInfo
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.plugins.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.math.*

object OsmRepository {

    private const val OVERPASS_URL = "https://overpass-api.de/api/interpreter"
    private const val LOOKAHEAD_DEG = 0.18   // ~20 km
    private const val TRACK_RADIUS_M = 200

    private val json = Json { ignoreUnknownKeys = true; coerceInputValues = true }

    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) { json(json, contentType = ContentType.Any) }
        install(HttpTimeout) {
            requestTimeoutMillis = 25_000
            connectTimeoutMillis = 5_000
            socketTimeoutMillis = 25_000
        }
    }

    suspend fun fetchTrackData(lat: Double, lon: Double): OsmTrackData = withContext(Dispatchers.IO) {
        try {
            val trackInfo = fetchTrackInfo(lat, lon)
            val features = fetchFeatures(lat, lon)
            OsmTrackData(trackInfo = trackInfo, features = features)
        } catch (e: Exception) {
            Log.e("OsmRepository", "fetchTrackData failed: ${e::class.simpleName} — ${e.message}")
            OsmTrackData(error = "Streckendaten konnten nicht geladen werden.")
        }
    }

    private suspend fun fetchTrackInfo(lat: Double, lon: Double): TrackInfo {
        val query = """
            [out:json][timeout:15];
            way(around:$TRACK_RADIUS_M,$lat,$lon)["railway"="rail"];
            out tags;
        """.trimIndent()
        return try {
            val response: OverpassResponse = client.get(OVERPASS_URL) {
                parameter("data", query)
            }.body()
            val tags = response.elements.firstOrNull()?.tags ?: return TrackInfo()
            TrackInfo(
                maxSpeed = tags["maxspeed"]?.toIntOrNull(),
                electrified = tags["electrified"],
                voltage = tags["voltage"]?.toIntOrNull(),
                tracks = tags["tracks"]?.toIntOrNull(),
                usage = tags["usage"]
            )
        } catch (e: Exception) {
            Log.e("OsmRepository", "fetchTrackInfo failed: ${e.message}")
            TrackInfo()
        }
    }

    private suspend fun fetchFeatures(lat: Double, lon: Double): List<RailFeature> {
        val south = lat - LOOKAHEAD_DEG
        val north = lat + LOOKAHEAD_DEG
        val west = lon - LOOKAHEAD_DEG
        val east = lon + LOOKAHEAD_DEG
        val bbox = "$south,$west,$north,$east"

        val query = """
            [out:json][timeout:20];
            (
              way["railway"="rail"]["tunnel"="yes"]($bbox);
              way["railway"="rail"]["bridge"="yes"]($bbox);
              node["railway"="station"]($bbox);
              node["railway"="halt"]($bbox);
            );
            out tags center;
        """.trimIndent()

        return try {
            val response: OverpassResponse = client.get(OVERPASS_URL) {
                parameter("data", query)
            }.body()

            response.elements.mapNotNull { element ->
                val elLat = element.lat ?: element.center?.lat ?: return@mapNotNull null
                val elLon = element.lon ?: element.center?.lon ?: return@mapNotNull null
                val tags = element.tags

                val type = when {
                    tags["railway"] == "station" -> RailFeatureType.STATION
                    tags["railway"] == "halt" -> RailFeatureType.HALT
                    tags["tunnel"] == "yes" -> RailFeatureType.TUNNEL
                    tags["bridge"] == "yes" -> RailFeatureType.BRIDGE
                    else -> return@mapNotNull null
                }

                val name = tags["name"] ?: tags["tunnel:name"] ?: tags["bridge:name"]

                // Skip unnamed bridges; show unnamed tunnels with generic label
                if (type == RailFeatureType.BRIDGE && name.isNullOrBlank()) return@mapNotNull null
                if ((type == RailFeatureType.STATION || type == RailFeatureType.HALT) && name.isNullOrBlank()) return@mapNotNull null

                RailFeature(
                    name = name ?: "Tunnel",
                    type = type,
                    distanceKm = haversineKm(lat, lon, elLat, elLon)
                )
            }
                .sortedBy { it.distanceKm }
                .distinctBy { it.type to it.name }
                .take(10)
        } catch (e: Exception) {
            Log.e("OsmRepository", "fetchFeatures failed: ${e.message}")
            emptyList()
        }
    }

    fun haversineKm(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val r = 6371.0
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val a = sin(dLat / 2).pow(2) +
                cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) * sin(dLon / 2).pow(2)
        return r * 2 * atan2(sqrt(a), sqrt(1 - a))
    }

    @Serializable
    private data class OverpassResponse(val elements: List<OverpassElement> = emptyList())

    @Serializable
    private data class OverpassElement(
        val type: String = "",
        val id: Long = 0L,
        val lat: Double? = null,
        val lon: Double? = null,
        val center: LatLon? = null,
        val tags: Map<String, String> = emptyMap()
    )

    @Serializable
    private data class LatLon(val lat: Double = 0.0, val lon: Double = 0.0)
}
